asset/file_reader/index.js

```diff
@@ -0,0 +1,108 @@
+'use strict';
+
+const Promise = require('bluebird');
+const Queue = require('@terascope/queue');
+const { getChunk, getOffsets } = require('@terascope/chunked-file-reader');
+const fse = require('fs-extra');
+
+function newSlicer(context, executionContext, retryData, logger) {
+    const opConfig = executionContext.config.operations[0];
+    const queue = new Queue();
+
+    async function processFile(filePath) {
+        const stat = await fse.stat(filePath);
+        const total = stat.size;
+        getOffsets(opConfig.size, total, opConfig.delimiter).forEach((offset) => {
+            offset.path = filePath;
+            queue.enqueue(offset);
+        });
+    }
+
+    async function getFilePaths(filePath) {
+        const dirContents = await fse.readdir(filePath);
+        return Promise.map(dirContents, async (file) => {
+            const fullPath = `${filePath}/${file}`;
+            const stats = await fse.lstat(fullPath);
+            if (stats.isFile()) {
+                return processFile(fullPath);
+            }
+            if (stats.isDirectory()) {
+                return getFilePaths(fullPath);
+            }
+            logger.error(`${file} is not a file or directory!!`);
+            return true;
+        })
+            .return([() => queue.dequeue()]);
```

Comment:
When running `once` jobs and a slicer returns a nil value, it will consider the slicer as complete. I would make sure that the queue was filled completely before allowing the slicer to return a nil value.

Reply:
Rejiggered the slicer based on your feedback to make it fully asynchronous and added the checks in its return to make sure it waits until valid slices are returned or there is an unplanned slicer shutdown.